kasa/smartprotocol.py

```diff
@@ -64,6 +73,22 @@ async def query(self, request: Union[str, Dict], retry_count: int = 3) -> Dict:
         """Query the device retrying for retry_count on failure."""
         async with self._query_lock:
             resp_dict = await self._query(request, retry_count)
+
+            if (
+                error_code := SmartErrorCode(resp_dict.get("error_code"))
+            ) != SmartErrorCode.SUCCESS:
+                msg = (
+                    f"Error querying device: {self.host}: "
+                    + f"{error_code.name}({error_code.value})"
+                )
+                if error_code in SMART_TIMEOUT_ERRORS:
+                    raise TimeoutException(msg)
+                if error_code in SMART_RETRYABLE_ERRORS:
+                    raise RetryableException(msg)
+                if error_code in SMART_AUTHENTICATION_ERRORS:
+                    raise AuthenticationException(msg)
+                raise SmartDeviceException(msg)
```

Comment:
Not probably something really necessary, but I was wondering if we should store the original error code / the error `SmartErrorCode` inside the exception object just in case someone wants to have a more specific handling for specific errors?

Reply:
I have a question on this suggestion but in the interest of not holding up getting this PR merged I will ask it on https://github.com/python-kasa/python-kasa/pull/569 and address it there if we decide.